### PR TITLE
Feature/shallow compare

### PR DESCRIPTION
**Pull request by michaelmerrill**
Shallow compare using the shallow obj function already in react-insula. Maybe we should pull these out into a utils folder that all packages can use? I also added tests for the cases that I could think of. Figure if we want to make this more performant the tests will help catch any regression.

### CODE REVIEW

File: packages/insula/src/ShallowCompare.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    return true;
+}
+
+export function shallowCompare(a, b) {
+    if (typeof a !== typeof b || Array.isArray(a) !== Array.isArray(b)) {
+        return false;
+    }
+
+    if (typeof a === 'function' || isPrimitive(a) || isPrimitive(b)) {
```
Comment (chandlerprall) [inline packages/insula/src/ShallowCompare.js]: I think...

```javascript
if (a === b) {
  return true;
} else if (typeof a === 'object' && typeof b === 'object') {
  return shallowObj(a, b);
} else {
  return false;
}
```

which catches primitive equality (first block), object/array shallow compare (second), and functions / anything else (last block). This doesn't account for other classes such as Map or Set, or Immutable's classes, or any other data structure as those `typeof === 'object'` but have no keys and are always "equal". Tempted to change the second `else if` to `else if(isArrayOrObject(a) && isArrayOrObject(b))` where

```javascript
function isArrayOrObject(x) {
  if (isPrimitive(x)) return false;
  return x.constructor === Object || x.constructor === Array;
}
```

Thoughts?

**Review by chandlerprall — CHANGES_REQUESTED** [review]

File: packages/insula/src/ShallowCompare.js (hunk 1 above)
Comment (michaelmerrill) [reply packages/insula/src/ShallowCompare.js]: That's much cleaner. There are a few edge cases with swapping out empty objects for empty arrays, but I can add a check in the shallowObj func that checks to see the constructors match.

**Comment by hartzis** [comment]
Heyo. did ya'll happen to discuss thoughts on just using other packages that do this?
facebooks - https://github.com/facebook/fbjs/blob/master/packages/fbjs/src/core/shallowEqual.js

or this guy with 2million dl's?

https://www.npmjs.com/package/shallowequal

**Comment by hartzis** [comment]
The reason i mention the other libraries is they would offer parity with what devs currently assume within the community since those are the "go to" solutions for shallow equal.

But if insula defines it's own it may differ in some slight ways and possibly cause confusion for some poor soul down the road.

**Comment by chandlerprall** [comment]
Ha. Both Michael and I looked and apparently didn't find the `shallowequal` module. Wow. Thanks Brian

**Comment by hartzis** [comment]
just a side note, i'm in no way saying that ideas/modules can't be improved upon, hence `insula` 😄 , but shallow compare in general feels pretty rock solid and keeping it consistent with other libraries feels like a better approach and not the focus of what `insula` is trying to solve.

**Comment by michaelmerrill** [comment]
I swear I looked for shallow equal packages! We should definitely use them. Plus, they look to be more optimized than my solution. Good catch @hartzis

**Pull request closed without merging**